### ISSUE TITLE
Add schedule-free adamw submission in JAX

### ISSUE THREAD

**Issue by priyakasimbeg** [mlcommons/algorithmic-efficiency#809]
## Description
Currently we have been unable to reproduce the schedule free adamw results with JAX. 
There seem to be differences between the optax implementation of schedule-free adamw and the pytorch submission.

**Comment by adefazio** [comment]
I can help debug any issues here. Do you have any code you can share? If there are issues with the optax jax implementation I want to get it fixed asap.

**Comment by adefazio** [comment]
There are many small differences between the behavior of schedule-free jax wrapper and the original algoperf submission. Some differences I'm aware of:

 - The bias correction in the submission scales the weight decay at early steps. This is slightly faster for fastMRI but doesn't appear to affect any other workloads in my experiments.
 - Weight decay is applied at y in the Jax version. This jax version is very similar in my experiments, if not slightly better. The experiments in the schedule-free paper use this decay-at-y version
 - There is a r=0.5 weighting in the submission version - this seems to make little if any difference in practice (hard to tell due to noise).

So overall I expect the jax wrapper version to give as good results on all problems (maybe slightly slower on fastmrI), so if there is a difference it would be from some sort of bug.

**Comment by priyakasimbeg** [comment]
Hi Aaron! thanks for weighing in on this. I seemed to have missed your messages on this thread. 

We have a slightly modified version based on the optax code here: https://github.com/priyakasimbeg/algorithmic-efficiency/blob/compare_schedule_free/tests/test_algorithms/schedule_free_adamw/jax/schedule_free_optax.py. 
This code adds r and we tested it with 0.75. 

I'm working on a test to compare the pytorch and jax implementations side by side but the test is not working yet. I can perhaps run a full training run on some of the workloads. 
But in the meantime feel free to weigh in again if you spot any other differences

**Comment by adefazio** [comment]
Ok, I take a look and see if I spot any differences.

**Comment by adefazio** [comment]
It looks like the z buffer my be initialized with zeros: 
https://github.com/priyakasimbeg/algorithmic-efficiency/blob/5556015054e3dda681e2a25e05a2f217d933453d/tests/test_algorithms/schedule_free_adamw/jax/submission.py#L58C51-L59C1
It needs to be initialized the same as the main parameter buffer. I think this line is a copy-paste error from the Jax version of NAdamW and other methods, where all optimizer state is normally initialized at zero.

Suggestion: you might want to set z on the first call to the main optimizer update, that's what we do in the pytorch version.

**Comment by adefazio** [comment]
@priyakasimbeg Let me know if that initialization issue was the problem.

**Comment by priyakasimbeg** [comment]
Hi Aaron thanks for spotting that! 
 We did the end-to-end training with an internal codebase using the optax implementation: https://github.com/google-deepmind/optax/blob/3ba9822c2a8d5fa7d046180b2574e108094523b4/optax/contrib/_schedule_free.py. It's not immediately obvious to me how this is initialized but will investigate and report back here

**Comment by init-22** [comment]
Hello @adefazio 
Just following up on this thread, me and @priyakasimbeg were actually debugging this issue and found that z buffer initialization to zeros might actually be the reason for this, instead now we are passing model_params as suggested and it seems to solve the problem...

This is loss the val loss of Jax vs Pytorch looked when we initialize the z buffer as the above lines
<img width="1189" alt="Image" src="https://github.com/user-attachments/assets/b441d2a5-5093-4b0b-9c48-3b5914dcd773" />

and This is how it looks now:

<img width="1188" alt="Image" src="https://github.com/user-attachments/assets/4f27f29a-7b4e-4e4d-a21d-6f5806c4d18a" />

FYI, We are still going to run some more tests on it and observe the difference

**Comment by adefazio** [comment]
Awesome, let me know if there are any other issues I can help debug.

**Comment by wyfEmma** [comment]
I ran serveral other workloads comparing the diff between the jax implementation with proper initialized z-buffer and the pytorch implementation. 

The result we see shows diff for different workloads, here's the plot bwteen global_step and log-scaled validation loss:

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/a728dc0e-5633-4d0d-8848-bddfe5eb6c71" />

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/af0daf94-b348-48f2-924e-f1f504e8524c" />

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/fc7146bd-2992-4f8e-aa22-d1bdbb248243" />

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/15e2de16-f5b5-403a-96fa-d61bb230abad" />

will look into it more to find where the diff comes from

**Comment by wyfEmma** [comment]
interestingly for imagenet pytorch is unstable:

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/f167924f-c0d3-4c88-898a-547fe3d61ccd" />

**Comment by wyfEmma** [comment]
we have a finding that in the previous implementation, y is used in both training and evaluation phase, after fixing, here's the latest result(I also removed the grad clipping for jax because pytorch doesn't have it):

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/f3427537-a241-498f-a1c9-a2887e3272e7" />

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/3702dce1-e79d-4f46-adf9-2f974fa0e8a0" />

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/c8a81f51-e403-401c-bc8b-77a83d23c1a0" />

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/8be83e7b-802e-450a-8266-6d75c3c8a9b7" />

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/fddb501a-e3d0-477e-be55-0af2368ee5e2" />

**Comment by wyfEmma** [comment]
after fixing the replication error and memory issue:

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/8fc03cce-23fb-48c3-9deb-0ddd1a06181e" />

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/cfa037c7-f114-403f-b37e-48393f7c61f1" />

**Comment by wyfEmma** [comment]
<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/dabeb08d-8c09-485d-9303-157cd8bcf03c" />

**Comment by wyfEmma** [comment]
<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/64566f00-5c74-43a5-a304-6d6956a930cb" />

**Comment by wyfEmma** [comment]
<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/0dd7fb25-28b8-4ff1-acf2-80a31b091de4" />

**Comment by wyfEmma** [comment]
<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/40cbaa4c-8da6-493c-ab3d-fda85e8a98c0" />

**Comment by wyfEmma** [comment]
ok for the previous results, librispeech deepspeech was using ctc_loss, it looks very much like a scaling issue within ctc_loss,  here's the plot for wer, it shows that the two curves are aligning

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/5e9c9d28-b8a4-497c-abe0-732bbac0012a" />

**Comment by wyfEmma** [comment]
to see why ctc diff between Jax and pytorch in librispeech deepspeech: https://github.com/mlcommons/algorithmic-efficiency/issues/924

**Comment by wyfEmma** [comment]
why jax performed better than pytorch in finewebedulm?

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/f620b58f-9499-4044-933a-943f4a97da3b" />

I noticed in pytorch, there's a hyperparam r that controls as steps increase, how much z got averaged into x. if r>0, with step increase, z will be taken into account more when calculating x. pytorch set it to be 0.75 and jax set it as 0(no such logic)

but after i align the r hyperparameter, and the batch size, the diff is still there: 

<img width="1800" height="1050" alt="Image" src="https://github.com/user-attachments/assets/aa4a4d2c-75e5-4e10-8a4f-7d4e73e8f65b" />
